docs/proposals/trial-custom-crd.md

```diff
@@ -0,0 +1,189 @@
+We have implemented new trial template design in Katib v1beta1 ([katib/pull#1202](https://github.com/kubeflow/katib/pull/1202)
+and [katib/pull#1215](https://github.com/kubeflow/katib/pull/1215)) to make
+experiments valid YAML and make Katib more Kubernetes native.
+After migrating to the new API, users still can run only [BatchJob](https://kubernetes.io/docs/concepts/workloads/controllers/job/),
+[TFJob](https://github.com/kubeflow/tf-operator) or [PyTorchJob](https://github.com/kubeflow/pytorch-operator) in trial job.
+If we want to support new CRD, we need to manually change Katib controller source code.
+
+This approach makes impossible to use other CRDs in trial template, even if they satisfy trial job design.
+The number of various Kubernetes CRDs grows significantly and many users would like to use them in Katib
+(e.g, [katib/issue#1081, support Argo Workflow](https://github.com/kubeflow/katib/issues/1081)).
+Another reason to design unified approach is that CRD controller can have `Go` package versions
+that Katib controller doesn't support (e.g, [katib/issue#1081](https://github.com/kubeflow/katib/issues/1081#issuecomment-635338276)).
+
+That is why we propose a new controller design to support custom CRD in trial job and make Katib usable for various Kubernetes resources.
+To make this possible, we are changing API, trial controller, job provider, mutation webhook, metrics collector.
+
+## Goals
+
+1. Allow dynamic watchers for the custom CRD.
+2. Inject Katib sidecar container on training pod.
+3. Indicate training container for metrics collector execution.
+4. Run metrics collector parser after all pod processes completion.
+5. Get succeeded condition of running CRD.
+6. Verify that `sidecar.istio.io/inject: false` label is added.
+
+## Non-Goals
+
+1. Inject Katib sidecar container on more than one pod simultaneously.
+2. Specify list of succeeded conditions for the custom CRD.
+3. Dynamically add new trial watcher for the custom CRD without Katib restart.
+
+## Implementation
+
+During implementation this feature, we should not brake current Katib controller logic.
+Also, we need to make sure that CI is stable and it does not block other Katib work tasks.
+After completion, we can clean-up redundant code.
+
+### API
+
+To achieve above goals, we introduce these `TrialTemplate` API changes.
+
+```go
+
+// TrialTemplate describes structure of Trial template
+type TrialTemplate struct {
+  // Retain indicates that Trial resources must be not cleanup
+  Retain bool `json:"retain,omitempty"`
+
+  // Source for Trial template (unstructured structure or config map)
+  TrialSource `json:",inline"`
+
+  // List of parameters that are used in Trial template
+  TrialParameters []TrialParameterSpec `json:"trialParameters,omitempty"`
+
+  // Label that determines if pod needs to be injected by Katib sidecar container
+  PrimaryPodLabel map[string]string `json:"primaryPodLabel,omitempty"`
```

Comment:
I am curious about how will these extra fields be filled ?

Reply:
In this design it will look like this:
```yaml
PrimaryPodLabel:
  "label-key": "label-value"
```
Not sure if it is the best design.
We can follow the same API as [`metricStrategies`](https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/common/v1beta1/common_types.go#L80-L83):

```go
. . .
PrimaryPodLabel *PrimaryPodLabel
. . .

type PrimaryPodLabel struct {
	Name  string `json:"name,omitempty"`
	Value string `json:"value,omitempty"`
}
```

Does it make sense, if we have can set 1 label currently?

WDYT @gaocegege @johnugeorge ?

Reply:
I think we should provide a map or a slice here.

Reply:
Map looks good to me, since it follow k8s way: https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L226-L231.